Sources/UrlFormEncoding/UrlFormDecoder.swift

```diff
@@ -0,0 +1,629 @@
+
+  public func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
+    let container = self.parsingStrategy.strategy(String(decoding: data, as: UTF8.self))
+    self.containers.append(container)
+    defer { self.containers.removeLast() }
+    return try T(from: self)
+  }
+
+  private func unbox(_ value: Any, as type: Data.Type) throws -> Data {
+    guard let string = singleton(value) else {
+      throw Error.decodingError("Expected string data, got \(value)", self.codingPath)
+    }
+
+    switch self.dataDecodingStrategy {
+    case .deferredToData:
+      return try Data(from: self)
+    case .base64:
+      guard let data = Data(base64Encoded: string) else {
+        throw Error.decodingError("Expected base64-encoded data, got \(string)", self.codingPath)
+      }
+      return data
+    case let .custom(strategy):
+      guard let data = strategy(string) else {
+        throw Error.decodingError("Failed strategy when decoding data from \(string)", self.codingPath)
+      }
+      return data
+    }
+  }
+
+  private func unbox(_ value: Any, as type: Date.Type) throws -> Date {
+    guard let string = singleton(value) else {
+      throw Error.decodingError("Expected string date, got \(value)", self.codingPath)
+    }
+
+    switch self.dateDecodingStrategy {
+    case .deferredToDate:
+      return try Date(from: self)
+    case .secondsSince1970:
+      guard let date = Double(string).map(Date.init(timeIntervalSince1970:)) else {
+        throw Error.decodingError("Expected seconds, got \(string)", self.codingPath)
+      }
+      return date
+    case .millisecondsSince1970:
+      guard let date = Double(string).map({ Date(timeIntervalSince1970: $0 / 1000) }) else {
+        throw Error.decodingError("Expected milliseconds, got \(string)", self.codingPath)
+      }
+      return date
+    case .iso8601:
+      let someDate = iso8601DateFormatter.date(from: string)
+        ?? iso8601DateFormatterWithoutMilliseconds.date(from: string)
+      guard let date = someDate else {
+        throw Error.decodingError("Expected ISO 8601 date string, got \(string)", self.codingPath)
+      }
+      return date
+    case let .formatted(formatter):
+      guard let date = formatter.date(from: string) else {
+        throw Error.decodingError("Expected \(formatter.dateFormat), got \(string)", self.codingPath)
+      }
+      return date
+    case let .custom(strategy):
+      guard let data = strategy(string) else {
+        throw Error.decodingError("Failed strategy when decoding data from \(string)", self.codingPath)
+      }
+      return data
+    }
+  }
+
+  private func unbox<T: Decodable>(_ value: Any, as type: T.Type) throws -> T {
+    if type == Data.self {
+      return try self.unbox(value, as: Data.self) as! T
+    } else if type == Date.self {
+      return try self.unbox(value, as: Date.self) as! T
+    } else {
+      return try T(from: self)
+    }
+  }
+
+  public func container<Key>(keyedBy type: Key.Type) throws
```

Comment:
Quick NB for folks new to decoder development: there are 3 sub-types ("containers") that help a decoder decode:

- "Keyed" containers decode raw data into things that have stringly-typed "keys" (structs and their named properties, dictionaries with string keys).
- "Unkeyed" containers decode arrays of values.
- "Single value" containers decode...a single value.

In this world of form data, we first parse our string into `[String: Any]`, handling arrays and nested dictionaries with whatever strategy our decoder picks. Once in this realm, we traverse the structure, handing each node off to the appropriate container implicitly:

- `[String: Any]` sub-nodes get sent to a "keyed" container.
- `[Any]` sub-nodes get sent to an "unkeyed" container.
- `Any`thing else gets sent to a single-value container.